kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -305,12 +326,27 @@ void inApiVersions(DecodedRequestFrame<ApiVersionsRequestData> apiVersionsFrame)
      * Called by the {@link ProxyChannelStateMachine} on entry to the {@link SelectingServer} state.
      */
     void inSelectingServer() {
-        // Pass this as the filter context, so that
-        // filter.initiateConnect() call's back on
-        // our initiateConnect() method
-        this.netFilter.selectServer(this);
-        this.proxyChannelStateMachine
-                .assertIsConnecting("NetFilter.selectServer() did not callback on NetFilterContext.initiateConnect(): filter='" + this.netFilter + "'");
+        List<FilterAndInvoker> apiVersionFilters = FilterAndInvoker.build("ApiVersionsIntersect (internal)", apiVersionsIntersectFilter);
+
+        NettyFilterContext filterContext = new NettyFilterContext(clientCtx.channel().eventLoop(), pfr);
+        List<FilterAndInvoker> filterChain = filterChainFactory.createFilters(filterContext, filters);
+        List<FilterAndInvoker> brokerAddressFilters = FilterAndInvoker.build("BrokerAddress (internal)",
+                new BrokerAddressFilter(endpointBinding.endpointGateway(), endpointReconciler));
+        var filters1 = new ArrayList<>(apiVersionFilters);
+        filters1.addAll(FilterAndInvoker.build("ApiVersionsDowngrade (internal)", apiVersionsDowngradeFilter));
+        filters1.addAll(filterChain);
+        if (endpointBinding.restrictUpstreamToMetadataDiscovery()) {
+            filters1.addAll(FilterAndInvoker.build("EagerMetadataLearner (internal)", new EagerMetadataLearner()));
+        }
+        filters1.addAll(brokerAddressFilters);
```

Comment:
```suggestion
        var filters1 = new ArrayList<>(apiVersionFilters);
        filters1.addAll(FilterAndInvoker.build("ApiVersionsDowngrade (internal)", apiVersionsDowngradeFilter));
        List<FilterAndInvoker> filterChain = filterChainFactory.createFilters(filterContext, filters);
        filters1.addAll(filterChain);
        if (endpointBinding.restrictUpstreamToMetadataDiscovery()) {
            filters1.addAll(FilterAndInvoker.build("EagerMetadataLearner (internal)", new EagerMetadataLearner()));
        }
        List<FilterAndInvoker> brokerAddressFilters = FilterAndInvoker.build("BrokerAddress (internal)",
                new BrokerAddressFilter(endpointBinding.endpointGateway(), endpointReconciler));
        filters1.addAll(brokerAddressFilters);
```
shuffle each bit closer to where it's installed in the list, could probably inline them too `filterChain` and `brokerAddressFilters` aren't doing much.